### PR TITLE
Add cross compilation support for Linux AArch64

### PR DESCRIPTION
**Pull request by odidev**
Closes https://github.com/pypa/cibuildwheel/issues/598

### CODE REVIEW

**Comment by odidev** [comment]
@joerick and @henryiii, could you please review the changes?

File: cibuildwheel/crosslinux.py [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+    BuildSelector,
+    NonPlatformWheelError,
+    get_build_verbosity_extra_flags,
+    prepare_command,
+    read_python_configs,
+)
+
+# Docker container which will be used to resove dependencies
+# while crosss compiling the wheels
+native_docker_images = {"aarch64": "quay.io/pypa/manylinux2014_aarch64"}
```
Comment (henryiii) [inline cibuildwheel/crosslinux.py:23]: This should be pulled from the normal setting for linux aarch64, not hardcoded. I'm assuming this also should be used for tests?

File: cibuildwheel/crosslinux.py [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+                    "run",
+                    "--rm",
+                    "--volume=/:/host",  # ignored on CircleCI
+                    native_docker_images[target_arch],
+                    "bash",
+                    "-c",
+                    target_arch_env.host_machine_tmp_in_container
+                    + '/install_deps.sh "'
+                    + cmd
+                    + '"',
```
Comment (henryiii) [inline cibuildwheel/crosslinux.py:112]:
```suggestion
                    f'{target_arch_env.host_machine_tmp_in_container}/install_deps.sh "{cmd}"',
```

File: cibuildwheel/crosslinux.py [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+
+            # The instaleld dependencies are in /tmp/install_deps on host machine.
+            # Copy them into the toolchain
+            dir_list = os.listdir(target_arch_env.host_machine_deps_usr_out_container)
+            for dir in dir_list:
+                docker.call(
+                    [
+                        "cp",
+                        "-rf",
+                        target_arch_env.host_machine_deps_usr_in_container + "/" + dir,
```
Comment (henryiii) [inline cibuildwheel/crosslinux.py:125]:
```suggestion
                        f"{target_arch_env.host_machine_deps_usr_in_container}/{dir}",
```

File: cibuildwheel/crosslinux.py [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+                    "-c",
+                    target_arch_env.host_machine_tmp_in_container
+                    + '/install_deps.sh "'
+                    + cmd
+                    + '"',
+                ],
+                check=True,
+            )
+
+            # The instaleld dependencies are in /tmp/install_deps on host machine.
```
Comment (henryiii) [inline cibuildwheel/crosslinux.py:117]:
```suggestion
            # The installed dependencies are in /tmp/install_deps on host machine.
```

File: cibuildwheel/crosslinux.py [hunk 5]
```diff
@@ -0,0 +1,536 @@
...
+                docker.call(
+                    [
+                        "cp",
+                        "-rf",
+                        target_arch_env.host_machine_deps_usr_in_container + "/" + dir,
+                        target_arch_env.toolchain_deps,
+                    ]
+                )
+        elif cmd.startswith("pip ") or cmd.startswith("python ") or cmd.startswith("python3 "):
+            if pip_install_env_create is True and before_build is True:
```
Comment (henryiii) [inline cibuildwheel/crosslinux.py:130]:
```suggestion
            if pip_install_env_create and before_build:
```

File: cibuildwheel/crosslinux.py [hunk 6]
```diff
@@ -0,0 +1,536 @@
...
+            docker.call(["sh", "-c", cmd], env=env)
+        else:
+            print(
+                "During cross compilation, in wheel build phase, only pip/python/yum related commands are allowed"
+            )
+            invalid_cmd = True
+            break
+    docker.call(["rm", "-rf", tmpdirpath])
+
+    if invalid_cmd is True:
```
Comment (henryiii) [inline cibuildwheel/crosslinux.py:164]:
```suggestion
    if invalid_cmd:
```

File: cibuildwheel/crosslinux.py [hunk 7]
```diff
@@ -0,0 +1,536 @@
...
+    target_arch_env = TargetArchEnvUtil(env.get("CROSS_ROOT"), target_arch)
+
+    cmds = [cmd.strip().replace("\t", " ") for cmd in cmd_str.split("&&")]
+
+    # Copy install_deps.sh script from container's tmp to host machine tmp and use it
+    if not os.path.isfile(target_arch_env.tmp + "/install_deps.sh"):
+        docker.call(
+            [
+                "cp",
+                target_arch_env.tmp + "/install_deps.sh",
```
Comment (henryiii) [inline cibuildwheel/crosslinux.py:83]:
```suggestion
                f"{target_arch_env.tmp}/install_deps.sh",
```

File: cibuildwheel/crosslinux.py [hunk 8]
```diff
@@ -0,0 +1,536 @@
...
+                    # We will repair the wheel in a different environment, copy the
+                    # built wheels back on host machine along with the script used to
+                    # repair the wheel
+                    docker.call(
+                        ["cp", "-r", temp_dir, target_arch_env.host_machine_tmp_in_container]
+                    )
+                    docker.call(
+                        [
+                            "cp",
+                            target_arch_env.tmp + "/repair_wheel.sh",
```
Comment (henryiii) [inline cibuildwheel/crosslinux.py:386]:
```suggestion
                            f"{target_arch_env.tmp}/repair_wheel.sh",
```

File: cibuildwheel/crosslinux.py [hunk 9]
```diff
@@ -0,0 +1,536 @@
...
+    get_build_verbosity_extra_flags,
+    prepare_command,
+    read_python_configs,
+)
+
+# Docker container which will be used to resove dependencies
+# while crosss compiling the wheels
+native_docker_images = {"aarch64": "quay.io/pypa/manylinux2014_aarch64"}
+
+# Cross triple for different architectures
```
Comment (henryiii) [inline cibuildwheel/crosslinux.py:25]: What is a triple? "cross_triple" is the name of this, so describing it as "Cross triple" isn't very helpful in a comment.

File: cibuildwheel/architecture.py [hunk 10]
```diff
@@ -80,6 +89,10 @@ def all_archs(platform: PlatformName) -> "Set[Architecture]":
                 Architecture.ppc64le,
                 Architecture.s390x,
             }
+        elif platform == "crosslinux":
+            return {
+                Architecture.aarch64,
+            }
```
Comment (henryiii) [inline cibuildwheel/architecture.py:95]:
```suggestion
            return {Architecture.aarch64}
```

File: cibuildwheel/__main__.py [hunk 11]
```diff
@@ -400,7 +404,7 @@ def get_build_identifiers(
         List[cibuildwheel.macos.PythonConfiguration],
     ]
 
-    if platform == "linux":
+    if platform == "linux" or platform == "crosslinux":
```
Comment (henryiii) [inline cibuildwheel/__main__.py:407]:
```suggestion
    if platform in {"linux", "crosslinux"}:
```

Minor, but CPython optimizes this so it's at least as fast as the other expression, and it reads better.

File: cibuildwheel/__main__.py [hunk 12]
```diff
@@ -278,7 +280,7 @@ def main() -> None:
         sys.exit(0)
 
     manylinux_images: Dict[str, str] = {}
-    if platform == "linux":
+    if platform == "linux" or platform == "crosslinux":
```
Comment (henryiii) [inline cibuildwheel/__main__.py:283]:
```suggestion
    if platform in {"linux", "crosslinux"}:
```

**Review by henryiii — COMMENTED** [review]
A few very minor things, but I'm generally in favor. @joerick?

We also need this to run in our testing matrix. I think it should run in the same places we run our emulated tests now, since it also requires emulation, and it should be faster (our tests don't compile that much, so might not see a huge difference here), so I think it could be fairly easily added alongside the existing runs.

I can try this out on a package (like boost-histogram) later.

**Comment by odidev** [comment]
@henryiii, I have pushed the changed. Can you please have a look. Thanks.

File: cibuildwheel/__main__.py (hunk 12 above)
Comment (odidev) [reply cibuildwheel/__main__.py]: Left as it is to satisfy the mypy.

File: cibuildwheel/__main__.py (hunk 11 above)
Comment (odidev) [reply cibuildwheel/__main__.py]: Left as it is to satisfy the mypy.

**Comment by henryiii** [comment]
Just a list of packages that I work on or help with where this would likely dramatically improve compile time:

* cmake-python-distributions (really important, because we currently have to disable building and testing except for releases, because emulation is too slow and Travis runs out of credits)
* ninja-python-distributions (ditto)
* boost-histogram
* iminuit
* awkward (and 2-3 other scikit-hep packages, but these three are the main ones)
* pybind11 examples (currently do not show how to build for other archs)

Once this is pretty close (that is, once we get a review from at least @joerick and/or @mayeut and/or @YannickJadoul), I'll try a couple of those and see if it works and what the performance is. Or sooner f I have a few minutes.

File: cibuildwheel/crosslinux.py [hunk 13]
```diff
@@ -0,0 +1,531 @@
...
+                    log.step("Building wheel...")
+
+                    temp_dir = PurePath("/tmp/cibuildwheel")
+                    built_wheel_dir = temp_dir / "built_wheel"
+                    docker.call(["rm", "-rf", built_wheel_dir])
+                    docker.call(["mkdir", "-p", built_wheel_dir])
+
+                    verbosity_flags = get_build_verbosity_extra_flags(options.build_verbosity)
+
+                    # pip wheel is not working properly with crossenv, use "bdist_wheel" for now
```
[the remark this answers is not in the document]
Comment (joerick) [reply cibuildwheel/crosslinux.py:331]: 👍

**Comment by joerick** [comment]
I'm having a tidy up of PRs, so I'm closing this for now. Feel free to reopen if you'd like to continue work on this approach.

**Pull request closed without merging**